doc/source/WorkingWithTheNodeGraph/Context/index.md

```diff
@@ -0,0 +1,331 @@
+
+All other Context Variables must be declared and initialized, either in the graph itself or in the Gaffer app that's executing the graph. They can handle all the same value types as plug, such as string, int, float, vector, matrix, etc. These other Context Variables are often used to pass an arbitrary variable value upstream.
+
+Many of the iterative nodes, such as Loop, Wedge, Random, CollectScenes, and CollectImages have a string plug that generates a Context Variable to vary upstream behaviour. Default names are typically provided to these plugs for convenience, but you can provide your own.
+
+Here are some examples of default names for Context Variables you will come across in these nodes:
+
+```eval_rst
+========================= =================== ============================================================
+Context Variable          Source node         Meaning
+========================= =================== ============================================================
+:code:`loop:index`        Loop                Current index of the loop.
+:code:`wedge:value`       Wedge               Value of the current wedge.
+:code:`collect:rootName`  CollectScenes       Name of the current root location being added to the scene.
+========================= =================== ============================================================
+```
+
+> Note: The `prefix:name` format in the default Context Variable names on has no special significance to the string algorithm.
+
+A Context Variable created at the root of the graph can be referred to as a global Context Variable. Global Context Variables are always passed into all Contexts, regardless of which node is being executed. They are declared and initialized in theVariables tab of the Settings menu, and can be overridden in individual networks.
+
+By default, two global Context Variables are added to all graphs:
+
+```eval_rst
+============================= ===================================================================
+Global Context Variable       Result
+============================= ===================================================================
+:code:`project:name`          The name of the project the graph belongs to.
+:code:`project:rootDirectory` The directory where the current graph's project files are stored.
+============================= ===================================================================
+```
+
+
+## Context flow ##
+
+While 2D and 3D data flows down the graph, Contexts somewhat counter-intuitively flow upstream, from child node to parent node. This is due to how Gaffer evaluates graphs. When traversing the graph to compute the final plug values, it starts from the calling node, then proceeds upward, passing along the Context if a plug (including a node's main out plug) must have a value computed. Most of the time this flow isn't noticeable, but it becomes pertinent for Context Variables and deep inspection of 2D and 3D data via the Python API.
```

Comment:
I tried to go with Tom's suggestion: https://docs.google.com/document/d/15mKlkKX3ApvGhgUodfnO6-5yaQ6YgkxS_ijR5Q-n1jQ/edit?disco=AAAAJUGV6Cg

Reply:
Regarding the stills, I think another illustration would be appropriate, here: https://docs.google.com/document/d/15mKlkKX3ApvGhgUodfnO6-5yaQ6YgkxS_ijR5Q-n1jQ/edit?disco=AAAAJUGV6Co. It would be extra work, but it's probably worth it.

Reply:
> I tried to go with Tom's suggestion: https://docs.google.com/document/d/15mKlkKX3ApvGhgUodfnO6-5yaQ6YgkxS_ijR5Q-n1jQ/edit?disco=AAAAJUGV6Cg

I think thats good introduction, thanks. Sorry I've made some suggested tweaks to the wording to try and better align with the ground truth whilst keeping your structure.

> Regarding the stills, I think another illustration would be appropriate, here: https://docs.google.com/document/d/15mKlkKX3ApvGhgUodfnO6-5yaQ6YgkxS_ijR5Q-n1jQ/edit?disco=AAAAJUGV6Co. It would be extra work, but it's probably worth it.

Do you mean to illustrate the variable flow as annotations to that one? That could work. I think it might need doctoring quite a bit though so its maybe more a composite than a screenshot as the expressions etc.. are a bit hard to read. I think it'd def be worth the effort. 

I guess my only real concern with that one though is that it introduces a lot of new concepts at this very early stage (ie: custom context variables being set half way up), before we've even cemented the basic premise that Contexts go up.

Perhaps a simple initial diagram that is as basic as switch node and some arrows showing the context going up somehow, as you did in the video perhaps? This would then illustrates the specific points were making in the into text about only queried plugs cause upstream compute, and that the Context is passed upwards?

Reply:
de816fa1978b494bb573b5c12e8377c869a2be77